### PR TITLE
Implement recovery behavior for LocalTrajectoryUploader

### PR DESCRIPTION
**Pull request by cschuet**

### CODE REVIEW

File: cartographer/cloud/internal/handlers/add_trajectory_handler.cc [hunk 1]
```diff
@@ -60,10 +60,15 @@ void AddTrajectoryHandler::OnRequest(
...
-    GetContext<MapBuilderContextInterface>()
-        ->local_trajectory_uploader()
-        ->AddTrajectory(request.client_id(), trajectory_id, expected_sensor_ids,
-                        trajectory_builder_options);
+    if (!GetContext<MapBuilderContextInterface>()
+             ->local_trajectory_uploader()
+             ->AddTrajectory(request.client_id(), trajectory_id,
+                             expected_sensor_ids, trajectory_builder_options)) {
+      LOG(ERROR) << "Failed to create trajectory in uplink: " << trajectory_id;
+      Finish(::grpc::Status(::grpc::INTERNAL,
```
Comment (gaschler) [inline cartographer/cloud/internal/handlers/add_trajectory_handler.cc]: Is this covered in the integration test?

File: cartographer/cloud/internal/local_trajectory_uploader.cc [hunk 2]
```diff
@@ -120,6 +144,41 @@ void LocalTrajectoryUploader::Shutdown() {
...
+    proto::SensorData* sensor_data =
+        send_queue_.PeekWithTimeout<proto::SensorData>(kPopTimeout);
+    if (sensor_data) {
+      CHECK_GE(sensor_data->local_slam_result_data().submaps_size(), 0);
+      if (sensor_data->sensor_data_case() ==
+              proto::SensorData::kLocalSlamResultData &&
+          sensor_data->local_slam_result_data()
+                  .submaps(
+                      sensor_data->local_slam_result_data().submaps_size() - 1)
+                  .submap_2d()
```
Comment (gaschler) [inline cartographer/cloud/internal/local_trajectory_uploader.cc]: Also handle 3D

File: cartographer/cloud/internal/local_trajectory_uploader.cc [hunk 3]
```diff
@@ -143,9 +202,17 @@ void LocalTrajectoryUploader::ProcessSendQueue() {
...
-        CHECK(client.Write(batch_request));
+            client_channel_,
+            async_grpc::CreateUnlimitedConstantDelayStrategy(
+                common::FromSeconds(1), kUnrecoverableStatusCodes));
+        LOG(INFO) << "Uploading";
+        if (!client.Write(batch_request)) {
+          // Unrecoverable error occurred.
+          batch_request.clear_sensor_data();
+          TryRecovery();
+        } else {
```
Comment (gaschler) [inline cartographer/cloud/internal/local_trajectory_uploader.cc]: "if (positive) ... continue; ..." may be more readable than "if (negative) ... else ..."

File: cartographer/cloud/internal/local_trajectory_uploader.cc [hunk 4]
```diff
@@ -154,15 +221,17 @@ void LocalTrajectoryUploader::ProcessSendQueue() {
...
+                                .uplink_trajectory_id;
   sensor_metadata->set_trajectory_id(cloud_trajectory_id);
 }
 
-void LocalTrajectoryUploader::AddTrajectory(
+bool LocalTrajectoryUploader::AddTrajectory(
     const std::string& client_id, int local_trajectory_id,
     const std::set<SensorId>& expected_sensor_ids,
     const mapping::proto::TrajectoryBuilderOptions& trajectory_options) {
+  LOG(INFO) << "LocalTrajectoryUploader::AddTrajectory " << client_id;
```
Comment (gaschler) [inline cartographer/cloud/internal/local_trajectory_uploader.cc]: add "client_id:" or drop line

File: cartographer/cloud/internal/local_trajectory_uploader.cc [hunk 5]
```diff
@@ -120,6 +144,41 @@ void LocalTrajectoryUploader::Shutdown() {
...
+      if (sensor_data->sensor_data_case() ==
+              proto::SensorData::kLocalSlamResultData &&
+          sensor_data->local_slam_result_data()
+                  .submaps(
+                      sensor_data->local_slam_result_data().submaps_size() - 1)
+                  .submap_2d()
+                  .num_range_data() == 1) {
+        break;
+      } else {
+        send_queue_.Pop();
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer/cloud/internal/local_trajectory_uploader.cc]: This is in the else branch of if(sensor_data) so Peek did not time out and there is something to pop.

File: cartographer/cloud/internal/handlers/add_trajectory_handler.cc (hunk 1 above)
Comment (cschuet) [reply cartographer/cloud/internal/handlers/add_trajectory_handler.cc]: this is slightly tricky. I will do this in another PR.

File: cartographer/cloud/internal/local_trajectory_uploader.cc (hunk 2 above)
Comment (cschuet) [reply cartographer/cloud/internal/local_trajectory_uploader.cc]: done

File: cartographer/cloud/internal/local_trajectory_uploader.cc (hunk 4 above)
Comment (cschuet) [reply cartographer/cloud/internal/local_trajectory_uploader.cc]: done

File: cartographer/cloud/internal/local_trajectory_uploader.h [hunk 6]
```diff
@@ -44,13 +44,12 @@ class LocalTrajectoryUploaderInterface {
   // Enqueue an Add*DataRequest message to be uploaded.
   virtual void EnqueueSensorData(
       std::unique_ptr<proto::SensorData> sensor_data) = 0;
-  virtual void AddTrajectory(
+  virtual bool AddTrajectory(
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer/cloud/internal/local_trajectory_uploader.h]: done

File: cartographer/cloud/internal/local_trajectory_uploader.cc (hunk 3 above)
Comment (cschuet) [reply cartographer/cloud/internal/local_trajectory_uploader.cc]: done

**Comment by cschuet** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user cschuet. Merge queue now has a length of 1.

**Pull request merged**